managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java

```diff
@@ -737,6 +738,69 @@ public void testBacklogStatsWhenDroppingData() throws Exception {
         ledger.close();
     }
 
+    @Test
+    public void testInvalidateReadHandleWithSlowNonDurableCursor() throws Exception {
+        ManagedLedgerImpl ledger = (ManagedLedgerImpl) factory.open("testInvalidateReadHandleWithSlowNonDurableCursor",
+                new ManagedLedgerConfig().setMaxEntriesPerLedger(1).setRetentionTime(-1, TimeUnit.SECONDS)
+                        .setRetentionSizeInMB(-1));
+        ManagedCursor c1 = ledger.openCursor("c1");
+        ManagedCursor nonDurableCursor = ledger.newNonDurableCursor(PositionImpl.EARLIEST);
+
+        assertEquals(nonDurableCursor.getNumberOfEntries(), 0);
+        assertEquals(nonDurableCursor.getNumberOfEntriesInBacklog(true), 0);
+
+        List<Position> positions = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            positions.add(ledger.addEntry(("entry-" + i).getBytes(UTF_8)));
+        }
+
+        CountDownLatch latch = new CountDownLatch(10);
+        for (int i = 0; i < 10; i++) {
+            ledger.asyncReadEntry((PositionImpl) positions.get(i), new AsyncCallbacks.ReadEntryCallback() {
+                @Override
+                public void readEntryComplete(Entry entry, Object ctx) {
+                    latch.countDown();
+                }
+
+                @Override
+                public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
+                    latch.countDown();
+                }
+            }, null);
+        }
+
+        latch.await();
+
+        ledger.updateTheSlowestNonDurableReadPosition(nonDurableCursor.getReadPosition());
+        c1.markDelete(positions.get(4));
+
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        ledger.internalTrimConsumedLedgers(promise);
+        promise.join();
+
+        Assert.assertTrue(ledger.ledgerCache.containsKey(3));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(4));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(5));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(6));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(7));
+
+        promise = new CompletableFuture<>();
+
+        nonDurableCursor.markDelete(positions.get(3));
+        ledger.updateTheSlowestNonDurableReadPosition(nonDurableCursor.getReadPosition());
+
+        ledger.internalTrimConsumedLedgers(promise);
+        promise.join();
+
+        Assert.assertFalse(ledger.ledgerCache.containsKey(3));
+        Assert.assertFalse(ledger.ledgerCache.containsKey(4));
+        Assert.assertFalse(ledger.ledgerCache.containsKey(5));
+        Assert.assertFalse(ledger.ledgerCache.containsKey(6));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(7));
+
```

Comment:
And we should continue to update the `SlowestNonDurableReadPosition` to make sure the read handle cache can be invalidated

Reply:
Here already invalidate ledger 3,4,5,6.

Reply:
> And we should continue to update the SlowestNonDurableReadPosition to make sure the read handle cache can be invalidated
I think @codelipenghui means to do repeatedly like line 789~790.

> Here already invalidate ledger 3,4,5,6.
Why invalidate `4,5,6` ?

Reply:
positions[0] is `3:0`,  positions[1] is `4:0`,  positions[2] is `5:0`, positions[3] is `6:0`, positions[4] `7:0`

At line 789,790. we make non-durable cursor mark-delete to 6:0, the next read postion is `7:0`, the managed-ledger non-durable slowest read position is `7:0`. So the ledger `3,4,5,6` will be invalidate

Reply:
Ah, thanks.